### PR TITLE
Add :sample_binary_id option for generators

### LINKED ISSUE #1527: Generators produce controller tests with ID's which fail against Mongo object ID's.
I asked about this on the mongo_ecto project, but it appears the matter is upstream in the nature of the Phoenix generator. Whatever the case, it causes the generated controller tests from pheonix.gen.html and phoenix.gen.json to fail.

I am using mongo_ecto, which talks to MongoDB and uses ObjectIDs as it ID type. If I run this command: 
```
mix phoenix.gen.html Car cars name:string description:string
```
the generated controller test file will contain this test: 
```
  test "renders page not found when id is nonexistent", %{conn: conn} do
    assert_error_sent 404, fn ->
      get conn, car_path(conn, :show, "11111111-1111-1111-1111-111111111111")
    end
  end
```
This test fails, because the ID string is invalid. If this ID is changed to this: 
```
  test "renders page not found when id is nonexistent", %{conn: conn} do
    assert_error_sent 404, fn ->
      get conn, car_path(conn, :show, "56a537b63e0eee155869a051")
    end
  end
```
Then the test succeeds. In the discussion on the mongo_ecto project, while the issue is specific to MongoDB's ID requirements, the answer I received is that because of the nature of the Phoenix generators, this is not something that can be fixed in the mongo_ecto code, but has to be fixed upstream. 

I confess I don't have the knowledge of the Phoenix or mongo_ecto codebase to speak further on the matter, but if someone can point me to the place in the Phoenix source which is in play here, I'd be happy to take a look and brainstorm on a solution. At any rate, it does constitute a bug, and some provision should be made to either fix it outright in Phoenix, or to expose functionality to allow it to be fixed downstream in mongo_ecto.

### PR DESCRIPTION
**Pull request by michalmuskala**
Closes  #1527

### CODE REVIEW

File: installer/lib/phoenix_new.ex [hunk 1]
```diff
@@ -408,7 +409,8 @@ defmodule Mix.Tasks.Phoenix.New do
       prod: [database: "db/#{app}_prod.sqlite"],
       test_begin: "Ecto.Adapters.SQL.begin_test_transaction(#{module}.Repo)",
       test_restart: "Ecto.Adapters.SQL.restart_test_transaction(#{module}.Repo, [])",
-      migration: true}
+      migration: true,
+      sample_binary_id: "11111111-1111-1111-1111-111111111111"}
```
Comment (josevalim) [inline installer/lib/phoenix_new.ex]: Let's not add those. If there is no sample_binary_id, we can't default to the one above.

File: priv/templates/phoenix.gen.json/controller_test.exs [hunk 2]
```diff
@@ -21,9 +21,9 @@ defmodule <%= module %>ControllerTest do
       "<%= k %>" => <%= singular %>.<%= k %><% end %>}
   end
 
-  test "does not show resource and instead throw error when id is nonexistent", %{conn: conn} do
+  test "renders page not found when id is nonexistent", %{conn: conn} do
```
Comment (josevalim) [inline priv/templates/phoenix.gen.json/controller_test.exs]: "raises with page not found" or "errors with page not found"?

File: test/mix/tasks/phoenix.gen.html_test.exs [hunk 3]
```diff
@@ -191,6 +194,19 @@ defmodule Mix.Tasks.Phoenix.Gen.HtmlTest do
...
+        Mix.Tasks.Phoenix.Gen.Json.run ["User", "users"]
+
+        assert_file "test/controllers/user_controller_test.exs", fn file ->
+          assert file =~ ~S|user_path(conn, :show, "abcd")|
+        end
+      end
+    end
+  end
+
+
```
Comment (josevalim) [inline test/mix/tasks/phoenix.gen.html_test.exs]: :scissors:

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

File: priv/templates/phoenix.gen.json/controller_test.exs (hunk 2 above)
Comment (michalmuskala) [inline priv/templates/phoenix.gen.json/controller_test.exs]: I made the message exactly the same as in `phoenix.gen.html` for this test. I can change it in the other place, or in both to something else.

Comment (josevalim) [inline priv/templates/phoenix.gen.json/controller_test.exs]: Ok, let's leave both as is then. :)

**Comment by chrismccord** [comment]
Can we get a a tiny doc in the phoenix_new moduledocs? Maybe even a "Third-pary Extension" section. Ready to go after that!

**Comment by josevalim** [comment]
Ping! :D

File: installer/lib/phoenix_new.ex (hunk 1 above)
Comment (michalmuskala) [inline installer/lib/phoenix_new.ex]: This general config is used for mysql, postgresql and mssql. Should I give this sample id explicitly in each adapter?

**Comment by michalmuskala** [comment]
@chrismccord I'm not sure phoenix_new is the best place to document those, because it's not configurable by the end user. On the other hand I can't find a plcace to document all the generator config values for phoenix.

File: installer/lib/phoenix_new.ex (hunk 1 above)
Comment (josevalim) [inline installer/lib/phoenix_new.ex]: No. We should only specify if if binary_id is given and it does not use UUID in its implementation.

**Comment by michalmuskala** [comment]
Updated

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**